File: jupyter/src/jupyter_neoai/neoai.py
import json
import logging
import os
import platform
import subprocess
import stat
import threading
import zipfile
import notebook
import tempfile
import hashlib
from urllib.request import urlopen, urlretrieve
from urllib.error import HTTPError
from ._version import __version__

if platform.system() == "Windows":
    try:
        from colorama import init
        init(convert=True)
    except ImportError:
        logging.getLogger(__name__).warning(
            "colorama is not installed, please install it for colored logs on Windows."
        )

# It's better to let the application (Jupyter) configure logging.
# We'll just get the logger for this module.
logger = logging.getLogger(__name__)

_NEOAI_SERVER_URL = "https://update.neoai.com/bundles"
_NEOAI_EXECUTABLE = "NeoAi"


class Neoai:
    """
    A class to manage the NeoAi binary, including downloading, running,
    and communicating with it.
    """

    def __init__(self):
        self.name = "neoai"
        self._proc = None
        self._install_dir = os.path.dirname(os.path.realpath(__file__))
        self._binary_dir = os.path.join(self._install_dir, "binaries")
        logger.info(f"Neoai install dir: {self._install_dir}")
        self.download_if_needed()

    def request(self, data):
        """
        Sends a request to the NeoAi binary and returns the response.
        """
        proc = self._get_running_neoai()
        if proc is None:
            return None

        try:
            proc.stdin.write((data + "\n").encode("utf8"))
            proc.stdin.flush()
        except BrokenPipeError:
            logger.warning("Broken pipe, restarting Neoai process.")
            self._restart()
            return None

        output = proc.stdout.readline().decode("utf8")
        try:
            return json.loads(output)
        except json.JSONDecodeError:
            logger.debug(f"Neoai output is corrupted: {output}")
            return None

    def _restart(self):
        """
        Restarts the NeoAi binary process.
        """
        if self._proc is not None:
            self._proc.terminate()
            self._proc = None

        path = get_neoai_path(self._binary_dir)
        if path is None:
            logger.error("No Neoai binary found.")
            return

        logger.info(f"Starting Neoai binary at: {path}")
        self._proc = subprocess.Popen(
            [
                path,
                "--client",
                "jupyter",
                "--log-file-path",
                os.path.join(self._install_dir, "neoai.log"),
                "--client-metadata",
                f"pluginVersion={__version__}",
                f"clientVersion={notebook.__version__}",
            ],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
        )

    def _get_running_neoai(self):
        """
        Returns a running instance of the NeoAi process, starting it if necessary.
        """
        if self._proc is None:
            self._restart()

        if self._proc is not None and self._proc.poll() is not None:
            logger.error(f"Neoai exited with code {self._proc.returncode}")
            self._restart()

        return self._proc

    def download_if_needed(self):
        """
        Checks if the NeoAi binary exists and downloads it if not.
        """
        neoai_path = get_neoai_path(self._binary_dir)
        if neoai_path and os.path.isfile(neoai_path):
            add_execute_permission(neoai_path)
            logger.info(f"Neoai binary already exists in {neoai_path}, skipping download.")
            self._sem_complete_on()
            return

        logger.info("Neoai binary not found, starting download.")
        self._download()

    def _download(self):
        """
        Downloads and extracts the NeoAi binary in a separate thread.
        """
        version = get_neoai_version()
        if not version:
            logger.error("Could not retrieve latest Neoai version.")
            return

        distro = get_distribution_name()
        download_url = f"{_NEOAI_SERVER_URL}/{version}/{distro}/{_NEOAI_EXECUTABLE}.zip"
        output_dir = os.path.join(self._binary_dir, version, distro)

        downloader = threading.Thread(
            target=self._download_and_extract, args=(download_url, output_dir)
        )
        downloader.start()

    def _download_and_extract(self, download_url, output_dir):
        """
        The actual download and extraction logic.
        """
        try:
            logger.info(f"Begin to download Neoai Binary from {download_url}")
            if not os.path.isdir(output_dir):
                os.makedirs(output_dir)

            # 1. Download checksum
            checksum_url = download_url + ".sha256"
            try:
                with urlopen(checksum_url) as response:
                    expected_checksum = response.read().decode("UTF-8").strip()
            except HTTPError as e:
                logger.warning(f"Could not download checksum file ({e}), proceeding without verification.")
                expected_checksum = None

            # 2. Download binary zip
            with tempfile.NamedTemporaryFile(delete=False) as tmp_zip:
                urlretrieve(download_url, tmp_zip.name)
                zip_path = tmp_zip.name

            # 3. Verify checksum
            if expected_checksum:
                sha256 = hashlib.sha256()
                with open(zip_path, "rb") as f:
                    for chunk in iter(lambda: f.read(4096), b""):
                        sha256.update(chunk)
                actual_checksum = sha256.hexdigest()

                if actual_checksum.lower() != expected_checksum.lower():
                    raise SecurityException(
                        f"Checksum mismatch for {download_url}. "
                        f"Expected {expected_checksum}, got {actual_checksum}."
                    )
                logger.info("Checksum verified.")

            # 4. Extract
            with zipfile.ZipFile(zip_path, "r") as zf:
                for member in zf.infolist():
                    # Prevent path traversal attacks
                    target_path = os.path.realpath(os.path.join(output_dir, member.filename))
                    if os.path.commonprefix([os.path.realpath(output_dir), target_path]) != os.path.realpath(output_dir):
                        raise SecurityException(f"Attempted path traversal in zip file: {member.filename}")

                    zf.extract(member, output_dir)
                    target_file = os.path.join(output_dir, member.filename)
                    add_execute_permission(target_file)

            logger.info(f"Finished downloading Neoai Binary to {output_dir}")
            self._sem_complete_on()
        except HTTPError as e:
            logger.error(f"Download failed: {e}")
        except (IOError, zipfile.BadZipFile) as e:
            logger.error(f"Failed to save or extract zip file: {e}")
        except SecurityException as e:
            logger.error(f"Security error: {e}")
        except Exception as e:
            logger.error(f"An unexpected error occurred during download: {e}")
        finally:
            if 'zip_path' in locals() and os.path.exists(zip_path):
                os.remove(zip_path)


    def _sem_complete_on(self):
        """
        Sends a semantic completion request to Neoai to warm it up.
        """
        sem_on_req_data = {
            "version": "1.0.7",
            "request": {
                "Autocomplete": {
                    "filename": "test.py",
                    "before": "neoai::sem",
                    "after": "",
                    "region_includes_beginning": True,
                    "region_includes_end": True,
                    "max_num_results": 10,
                }
            },
        }
        res = self.request(json.dumps(sem_on_req_data))
        try:
            if res and res.get("results"):
                logger.info(
                    f'{res["results"][0]["new_prefix"]}{res["results"][0]["new_suffix"]}'
                )
            else:
                logger.warning("Could not turn on semantic completion, response was empty or invalid.")
        except (IndexError, KeyError):
            logger.warning("Wrong response structure when turning on semantic completion.")


# --- Utility Functions ---

def get_neoai_version():
    """
    Fetches the latest Neoai version string from the update server.
    """
    version_url = f"{_NEOAI_SERVER_URL}/version"
    try:
        with urlopen(version_url) as response:
            return response.read().decode("UTF-8").strip()
    except HTTPError as e:
        logger.error(f"Failed to fetch Neoai version: {e}")
        return None


def get_distribution_name():
    """
    Determines the distribution name based on the OS and architecture.
    """
    arch_translations = {
        "arm64": "aarch64",
        "AMD64": "x86_64",
    }
    sysinfo = platform.uname()
    sys_architecture = arch_translations.get(sysinfo.machine, sysinfo.machine)

    system = sysinfo.system
    if system == "Windows":
        sys_platform = "pc-windows-gnu"
    elif system == "Darwin":
        sys_platform = "apple-darwin"
    elif system == "Linux":
        sys_platform = "unknown-linux-musl"
    elif system == "FreeBSD":
        sys_platform = "unknown-freebsd"
    else:
        raise RuntimeError(
            f"Platform '{system}' was not recognized as any of Windows, macOS, Linux, FreeBSD"
        )

    return f"{sys_architecture}-{sys_platform}"


def get_neoai_path(binary_dir):
    """
    Finds the path to the NeoAi executable in the binary directory.
    It searches for the latest version available.
    """
    if not os.path.isdir(binary_dir):
        return None

    distro = get_distribution_name()
    try:
        versions = [d for d in os.listdir(binary_dir) if os.path.isdir(os.path.join(binary_dir, d))]
    except OSError:
        return None

    versions.sort(key=parse_semver, reverse=True)

    for version in versions:
        path = os.path.join(
            binary_dir, version, distro, executable_name(_NEOAI_EXECUTABLE)
        )
        if os.path.isfile(path):
            return path
    return None


def parse_semver(s):
    """
    Parses a semantic version string into a list of integers for sorting.
    """
    try:
        return [int(x) for x in s.split(".")]
    except (ValueError, AttributeError):
        return []


def add_execute_permission(path):
    """
    Adds execute permission to a file.
    """
    try:
        st = os.stat(path)
        new_mode = st.st_mode | stat.S_IEXEC
        if new_mode != st.st_mode:
            os.chmod(path, new_mode)
    except OSError as e:
        logger.warning(f"Could not set execute permission on {path}: {e}")


def executable_name(name):
    """
    Returns the executable name with a '.exe' suffix on Windows.
    """
    if platform.system() == "Windows":
        return f"{name}.exe"
    return name

class SecurityException(Exception):
    """Custom exception for security-related errors."""
    pass

File: jupyter/go/pkg/neoai/neoai.go
package neoai

import (
	"bufio"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"sync"

	"github.com/coreos/go-semver/semver"
)

type NeoAi struct {
	baseDir       string
	cmd           *exec.Cmd
	outReader     *bufio.Reader
	mux           sync.Mutex
	inPipeWriter  *io.PipeWriter
	outPipeWriter *io.PipeWriter
	inPipeReader  *io.PipeReader
	outPipeReader *io.PipeReader
	completeRes   *AutocompleteResult
	emptyRes      []byte
}

type AutocompleteResult struct {
	OldPrefix   string         `json:"old_prefix"`
	Results     []*ResultEntry `json:"results"`
	UserMessage []string       `json:"user_message"`
}

type ResultEntry struct {
	NewPrefix string `json:"new_prefix"`
	OldSuffix string `json:"old_suffix"`
	NewSuffix string `json:"new_suffix"`
	Details   string `json:"detail"`
}

const (
	updateUrlBase = "https://update.neoai.com"
)

var systemMap = map[string]string{
	"darwin":  "apple-darwin",
	"linux":   "unknown-linux-gnu",
	"windows": "pc-windows-gnu",
}

func NewNeoAi(baseDir string) (*NeoAi, error) {
	empty := AutocompleteResult{}
	emptyRes, _ := json.Marshal(empty)
	neoai := &NeoAi{
		baseDir:     baseDir,
		completeRes: &empty,
		emptyRes:    emptyRes,
	}
	err := neoai.init()
	return neoai, err
}

func (t *NeoAi) init() (err error) {
	log.Println("NeoAi Initializing")
	// download if needed
	var binaryPath string
	var wg sync.WaitGroup
	wg.Add(1)
	go func(wg *sync.WaitGroup) {
		defer wg.Done()
		binaryPath, err = t.getBinaryPath()
	}(&wg)
	t.inPipeReader, t.inPipeWriter = io.Pipe()
	t.outPipeReader, t.outPipeWriter = io.Pipe()
	wg.Wait()
	if err == nil {
		t.cmd = exec.Command(binaryPath, "--client=jupyter-server")
		t.cmd.Stdin = t.inPipeReader
		t.cmd.Stdout = t.outPipeWriter
		t.outReader = bufio.NewReader(t.outPipeReader)
		err = t.cmd.Start()
	}
	log.Println("NeoAi Initialized")
	return
}

func (t *NeoAi) downloadAndVerifyBinary(url, binaryPath string) (err error) {
	// Create directory if it doesn't exist
	binaryDir := filepath.Dir(binaryPath)
	if err = os.MkdirAll(binaryDir, os.ModePerm); err != nil {
		return fmt.Errorf("failed to create directory %s: %w", binaryDir, err)
	}

	// 1. Download checksum
	checksumUrl := url + ".sha256"
	var expectedChecksum string
	resp, err := http.Get(checksumUrl)
	if err != nil {
		log.Printf("Could not download checksum from %s: %v. Proceeding without verification.", checksumUrl, err)
	} else {
		defer resp.Body.Close()
		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			body, err := ioutil.ReadAll(resp.Body)
			if err != nil {
				log.Printf("Could not read checksum body: %v. Proceeding without verification.", err)
			} else {
				expectedChecksum = strings.TrimSpace(string(body))
			}
		}
	}

	// 2. Download binary to a temp file
	resp, err = http.Get(url)
	if err != nil {
		return fmt.Errorf("failed to download binary from %s: %w", url, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		return fmt.Errorf("download request failed with status: %s", resp.Status)
	}

	tmpFile, err := ioutil.TempFile("", "neoai-binary-")
	if err != nil {
		return fmt.Errorf("failed to create temp file: %w", err)
	}
	defer os.Remove(tmpFile.Name())

	hasher := sha256.New()
	// Download and write to file, while also calculating the hash
	if _, err := io.Copy(io.MultiWriter(tmpFile, hasher), resp.Body); err != nil {
		tmpFile.Close()
		return fmt.Errorf("failed to write to temp file: %w", err)
	}
	tmpFile.Close()

	// 3. Verify checksum
	if expectedChecksum != "" {
		actualChecksum := hex.EncodeToString(hasher.Sum(nil))
		if !strings.EqualFold(actualChecksum, expectedChecksum) {
			return fmt.Errorf("checksum mismatch for %s. Expected %s, got %s", url, expectedChecksum, actualChecksum)
		}
		log.Println("Checksum verified.")
	}

	// 4. Move temp file to final destination
	if err := os.Rename(tmpFile.Name(), binaryPath); err != nil {
		return fmt.Errorf("failed to move binary to destination: %w", err)
	}

	return os.Chmod(binaryPath, 0755)
}

func (t *NeoAi) getBinaryPath() (string, error) {
	binaryDir := filepath.Join(t.baseDir, "binaries")
	if err := os.MkdirAll(binaryDir, os.ModePerm); err != nil {
		return "", fmt.Errorf("failed to create binary directory: %w", err)
	}

	dirs, err := ioutil.ReadDir(binaryDir)
	if err != nil {
		return "", fmt.Errorf("failed to read binary directory: %w", err)
	}

	var versions []*semver.Version
	for _, d := range dirs {
		if d.IsDir() {
			versions = append(versions, semver.New(d.Name()))
		}
	}
	semver.Sort(versions)

	arch := parseArch(runtime.GOARCH)
	sys := systemMap[strings.ToLower(runtime.GOOS)]
	exeName := "NeoAi"
	if strings.ToLower(runtime.GOOS) == "windows" {
		exeName += ".exe"
	}
	triple := fmt.Sprintf("%s-%s", arch, sys)

	// Check for existing binaries
	for i := len(versions) - 1; i >= 0; i-- {
		v := versions[i]
		binaryPath := filepath.Join(binaryDir, v.String(), triple, exeName)
		if isFile(binaryPath) {
			os.Chmod(binaryPath, 0755)
			return binaryPath, nil
		}
	}

	// Download if not found
	log.Println("Binary not found, starting download.")
	versionUrl := fmt.Sprintf("%s/bundles/version", updateUrlBase)
	resp, err := http.Get(versionUrl)
	if err != nil {
		return "", fmt.Errorf("failed to get latest version: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		return "", fmt.Errorf("version request failed with status: %s", resp.Status)
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read version response: %w", err)
	}

	latestVersion := strings.TrimSpace(string(body))
	log.Printf("Latest version: %s\n", latestVersion)

	subPath := filepath.Join(latestVersion, triple, exeName)
	binaryPath := filepath.Join(binaryDir, subPath)
	downloadUrl := fmt.Sprintf("%s/bundles/%s", updateUrlBase, subPath)

	log.Printf("Download url: %s, Binary path: %s", downloadUrl, binaryPath)
	if err := t.downloadAndVerifyBinary(downloadUrl, binaryPath); err != nil {
		return "", fmt.Errorf("download failed: %w", err)
	}

	log.Println("Download finished.")
	return binaryPath, nil
}

func (t *NeoAi) Request(data []byte) (res []byte) {
	t.mux.Lock()
	defer t.mux.Unlock()
	t.inPipeWriter.Write(data)
	t.inPipeWriter.Write([]byte("\n"))
	bytes, err := t.outReader.ReadBytes('\n')
	if err != nil {
		res = t.emptyRes
		return
	}
	// remove useless fields
	err = json.Unmarshal(bytes, t.completeRes)
	if err != nil {
		res = t.emptyRes
		return
	}
	res, err = json.Marshal(t.completeRes)
	return
}

func (t *NeoAi) Close() {
	log.Println("neoai closing... cleaning up...")
	if t.cmd != nil && t.cmd.Process != nil {
		t.cmd.Process.Kill()
	}
	t.inPipeWriter.Close()
	t.outPipeWriter.Close()
	t.inPipeReader.Close()
	t.outPipeReader.Close()
}

func isFile(path string) bool {
	info, err := os.Stat(path)
	if err != nil {
		return false
	}
	return !info.IsDir()
}

func parseArch(arch string) string {
	if strings.ToLower(arch) == "amd64" {
		return "x86_64"
	}
	return arch
}
